sw/device/silicon_creator/lib/otbn_util.c

```diff
@@ -0,0 +1,124 @@
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+#include "sw/device/silicon_creator/lib/drivers/otbn.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+void otbn_init(otbn_t *ctx) {
+  *ctx = (otbn_t){
+      .app = {0},
+      .app_is_loaded = false,
+      .error_bits = kOtbnErrBitsNoError,
+  };
+}
+
+uint32_t otbn_get_internal_err_bits(otbn_t *ctx) { return ctx->error_bits; }
+
+rom_error_t otbn_func_ptr_to_imem_addr(const otbn_t *ctx, otbn_ptr_t ptr,
+                                       uint32_t *imem_addr_otbn) {
+  const uint32_t *app_imem_start_addr = ctx->app.imem_start;
+  const uint32_t *app_imem_end_addr = ctx->app.imem_end;
+
+  if (ptr < app_imem_start_addr || ptr > app_imem_end_addr) {
+    return kErrorOtbnInvalidArgument;
+  }
+  *imem_addr_otbn = ptr - app_imem_start_addr;
+  return kErrorOk;
+}
+
+rom_error_t otbn_data_ptr_to_dmem_addr(const otbn_t *ctx, otbn_ptr_t ptr,
+                                       uint32_t *dmem_addr_otbn) {
+  if (ptr < ctx->app.dmem_start || ptr > ctx->app.dmem_end) {
+    return kErrorOtbnInvalidArgument;
+  }
+  *dmem_addr_otbn = ptr - ctx->app.dmem_start;
+  return kErrorOk;
+}
+
+rom_error_t otbn_busy_wait_for_done(otbn_t *ctx) {
+  while (otbn_is_busy()) {
+  }
+
+  otbn_err_bits_t err_bits;
+  otbn_get_err_bits(&err_bits);
+  if (err_bits != kOtbnErrBitsNoError) {
+    ctx->error_bits = err_bits;
+    return kErrorOtbnExecutionFailed;
+  }
+  return kErrorOk;
+}
+
+rom_error_t otbn_load_app(otbn_t *ctx, const otbn_app_t app) {
+  if (app.imem_end <= app.imem_start || app.dmem_end < app.dmem_start) {
+    return kErrorOtbnInvalidArgument;
+  }
+
+  const size_t imem_size = app.imem_end - app.imem_start;
+  const size_t dmem_size = app.dmem_end - app.dmem_start;
+
+  RETURN_IF_ERROR(otbn_imem_write(0, app.imem_start, imem_size));
+
+  if (dmem_size > 0) {
+    RETURN_IF_ERROR(otbn_dmem_write(0, app.dmem_start, dmem_size));
+  }
+
+  ctx->app = app;
+  ctx->app_is_loaded = true;
+  return kErrorOk;
+}
+
+rom_error_t otbn_execute_app(otbn_t *ctx) {
+  if (!ctx->app_is_loaded) {
+    return kErrorOtbnInvalidArgument;
+  }
+
+  RETURN_IF_ERROR(otbn_execute());
+
+  return kErrorOk;
+}
+
+rom_error_t otbn_copy_data_to_otbn(otbn_t *ctx, size_t len, const void *src,
+                                   otbn_ptr_t dest) {
+  uint32_t dest_dmem_addr;
+  RETURN_IF_ERROR(otbn_data_ptr_to_dmem_addr(ctx, dest, &dest_dmem_addr));
+  RETURN_IF_ERROR(otbn_dmem_write(dest_dmem_addr, src, len));
+
+  return kErrorOk;
+}
+
+rom_error_t otbn_copy_data_from_otbn(otbn_t *ctx, size_t len_bytes,
+                                     otbn_ptr_t src, void *dest) {
+  uint32_t src_dmem_addr;
+  RETURN_IF_ERROR(otbn_data_ptr_to_dmem_addr(ctx, src, &src_dmem_addr));
+  RETURN_IF_ERROR(otbn_dmem_read(src_dmem_addr, dest, len_bytes));
+
+  return kErrorOk;
+}
+
+rom_error_t otbn_zero_data_memory(otbn_t *ctx) {
```

Comment:
`ctx` is unused. This function can probably be moved to `otbn.c` and use `abs_mmio_write32` instead of going through `otbn_dmem_write`. I don't think there is a reason for this function to return anything since it shouldn't fail.